### PR TITLE
[REVIEW] Add BlazingSQL version of query 18

### PR DESCRIPTION
**Pull request by Christian8491**
This PR adds the version of "query" 18 using `BlazingSQL`.

To execute this "query" is needed some util functions from `tpcx_bb_query_18.py` file as  `find_relevant_reviews()`, `find_targets_in_reviews_helper()`, `create_found_reshaped_with_global_pos()`
Additionally the `negativeSentiment.txt` file

### CODE REVIEW

**Comment by Christian8491** [comment]
It shows **Correctness Assertion True** at scale factor **1K**, using **dgx201**

```
(tpcx-bb) christianc@rl-dgx2-d17-u05-rapids-dgx201:~/tpcx-bb/tpcx_bb/queries/q18$ python tpcx_bb_query_18_sql.py --data_dir=/raid/tpcx-bb/sf1000/parquet_2gb/ --verify_results --verify_dir=/mnt/weka/nicholasb/tpcx-bb/validation/sf1000/ --cluster_host=10.150.162.27
Using default arguments
Connected to tcp://10.150.162.27:8786
WARNING: Count not make directory
WARNING:root:WARNING: Count not make directory
WARNING: Count not make directory
WARNING:root:WARNING: Count not make directory
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/home/christianc/miniconda3/envs/tpcx-bb/lib/blazingsql-algebra.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
BlazingContext ready
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:722: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:722: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py:720: RuntimeWarning: Multiple sorted columns found ['s_name', 'sentiment'], cannot
 autodetect index. Will continue without an index.
To pick an index column, use the index= keyword; to 
silence this warning use index=False.
  RuntimeWarning,
Sentiment Analysis Query
95.8056992098885% overlap with 1533581 rows (RAPIDS denominator)
95.90632204766145% overlap with 1531972 rows (Spark denominator)
Correctness Assertion True
Must pass a sheet and tab name to use Google Sheets automation
```

The "query" finishes successfully at scale factor **10K**

```
(tpcx-bb) christianc@rl-dgx2-d17-u05-rapids-dgx201:~/tpcx-bb/tpcx_bb/queries/q18$ python tpcx_bb_query_18_sql.py --data_dir=/mnt/weka/tpcx-bb/sf10000/parquet_3gb/ --cluster_host=10.150.162.27
Using default arguments
Connected to tcp://10.150.162.27:8786
WARNING: Count not make directory
WARNING:root:WARNING: Count not make directory
WARNING: Count not make directory
WARNING:root:WARNING: Count not make directory
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/home/christianc/miniconda3/envs/tpcx-bb/lib/blazingsql-algebra.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
BlazingContext ready
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:722: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:722: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
Must pass a sheet and tab name to use Google Sheets automation
```

We need to consider how use these util functions  before change this PR to  **REVIEW**.

**Comment by beckernick** [comment]
Thanks for porting this and for the summary.

> We need to consider how use these util functions before change this PR to REVIEW.

You're 100% right. This will work single-node as is, but will fail multi-node. The short-term plan is:
- For any BSQL query that needs functions from the equivalent pure-Dask query, duplicate the function definitions in the BSQL query

Could you update this query (and others like it) to duplicate the necessary functions.

Eventually, we may want to factor them out and include them in the pip installable module. But for now, duplication is the path of least resistance.

**Comment by Christian8491** [comment]
After adding util functions from `tpcx_bb_query_18.py` file and the separate values as Spark version,  I  tested again at sf **1K** and got the **Correctness Assertion True message** using the **dgx201**, so updating this PR to **REVIEW**.

```
(tpcx-bb) christianc@rl-dgx2-d17-u05-rapids-dgx201:~/tpcx-bb/tpcx_bb/queries/q18$ python tpcx_bb_query_18_sql.py --data_dir=/raid/tpcx-bb/sf1000/parquet_2gb/ --verify_results --verify_dir=/mnt/weka/nicholasb/tpcx-bb/validation/sf1000/ --cluster_host=10.150.162.27
Using default arguments
Connected to tcp://10.150.162.27:8786
WARNING: Count not make directory
WARNING:root:WARNING: Count not make directory
WARNING: Count not make directory
WARNING:root:WARNING: Count not make directory
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/home/christianc/miniconda3/envs/tpcx-bb/lib/blazingsql-algebra.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
BlazingContext ready
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:722: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:722: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py:720: RuntimeWarning: Multiple sorted columns found ['s_name', 'sentiment'], cannot
 autodetect index. Will continue without an index.
To pick an index column, use the index= keyword; to 
silence this warning use index=False.
  RuntimeWarning,
Sentiment Analysis Query
95.8190700595487% overlap with 1533367 rows (RAPIDS denominator)
95.90632204766145% overlap with 1531972 rows (Spark denominator)
Correctness Assertion True
Must pass a sheet and tab name to use Google Sheets automation
```

**Comment by Christian8491** [comment]
Updated and tested again. 

1. It shows **Correctness Assertion True** at scale factor **1K**, using **dgx201**.

```
(tpcx-bb) christianc@rl-dgx2-d17-u05-rapids-dgx201:~/tpcx-bb/tpcx_bb/queries/q18$ python tpcx_bb_query_18_sql.py --config_file=/home/christianc/yaml_dir/validation_1k_config.yaml 
Using default arguments
Using default arguments
Connected to tcp://10.150.162.27:8786
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/home/christianc/miniconda3/envs/tpcx-bb/lib/blazingsql-algebra.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
BlazingContext ready
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:726: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:726: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py:720: RuntimeWarning: Multiple sorted columns found ['s_name', 'sentiment'], cannot
 autodetect index. Will continue without an index.
To pick an index column, use the index= keyword; to 
silence this warning use index=False.
  RuntimeWarning,
Sentiment Analysis Query
95.80851052730421% overlap with 1533536 rows (RAPIDS denominator)
95.90632204766145% overlap with 1531972 rows (Spark denominator)
Correctness Assertion True
Please install gspread and oauth2client               to use Google Sheets automation
```

from `validation_1k_config.yaml`, some useful values:

```
data_dir: /raid/tpcx-bb/sf1000/parquet_2gb/
verify_results: True
verify_dir: /mnt/weka/nicholasb/tpcx-bb/validation/sf1000/
```

2. The "query" finishes successfully at scale factor **10K**

```
(tpcx-bb) christianc@rl-dgx2-d17-u05-rapids-dgx201:~/tpcx-bb/tpcx_bb/queries/q18$ python tpcx_bb_query_18_sql.py --config_file=/home/christianc/yaml_dir/succes_10k_config.yaml 
Using default arguments
Using default arguments
Connected to tcp://10.150.162.27:8786
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/home/christianc/miniconda3/envs/tpcx-bb/lib/blazingsql-algebra.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
BlazingContext ready
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:726: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx-bb/lib/python3.7/site-packages/cudf/core/column/string.py:726: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
Please install gspread and oauth2client               to use Google Sheets automation
```

from `succes_10k_config.yaml`, some useful values:

```
data_dir: /mnt/weka/tpcx-bb/sf10000/parquet_3gb/
verify_results: False
verify_dir: 
```

File: tpcx_bb/queries/q18/tpcx_bb_query_18_sql.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import sys
+
+from xbb_tools.cluster_startup import attach_to_cluster
+import os
+import cudf
```
Comment (beckernick) [inline tpcx_bb/queries/q18/tpcx_bb_query_18_sql.py:22]: Can we remove this cudf import from the global scope and push this import into the `main` function?

Comment (Christian8491) [reply tpcx_bb/queries/q18/tpcx_bb_query_18_sql.py]: Sure. Updated

**Comment by Christian8491** [comment]
Updated and tested again at sf **1K** on the **dgx201**. It shows **Correctness Assertion True** (using the `negativeSentiment.txt` file just to test)

```
(tpcx) christianc@rl-dgx2-d17-u05-rapids-dgx201:~/tpcx-bb/tpcx_bb/queries/q18$ python tpcx_bb_query_18_sql.py --config_file=/home/christianc/yaml_dir/validation_1k_config.yaml
Using default arguments
Connected to tcp://10.150.162.27:8786
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/home/christianc/miniconda3/envs/tpcx/lib/blazingsql-algebra.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
BlazingContext ready
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/cudf/core/column/string.py:683: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/cudf/core/column/string.py:683: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py:720: RuntimeWarning: Multiple sorted columns found ['s_name', 'sentiment'], cannot
 autodetect index. Will continue without an index.
To pick an index column, use the index= keyword; to 
silence this warning use index=False.
  RuntimeWarning,
Sentiment Analysis Query
95.80851052730421% overlap with 1533536 rows (RAPIDS denominator)
95.90632204766145% overlap with 1531972 rows (Spark denominator)
Correctness Assertion True
Please install gspread and oauth2client               to use Google Sheets automation
```

**Comment by Christian8491** [comment]
Updated and tested again at sf **1K** on the **dgx201**. It shows **Correctness Assertion True** (using the `negativeSentiment.txt` file just to test)

```
(tpcx) christianc@rl-dgx2-d17-u05-rapids-dgx201:~/tpcx-bb/tpcx_bb/queries/q18$ python tpcx_bb_query_18_sql.py --config_file=/home/christianc/yaml_dir/validation_1k_config.yaml
Using default arguments
Connected to tcp://10.150.162.27:8786
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/home/christianc/miniconda3/envs/tpcx/lib/blazingsql-algebra.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
BlazingContext ready
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/cudf/core/column/string.py:683: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/cudf/core/column/string.py:683: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py:720: RuntimeWarning: Multiple sorted columns found ['s_name', 'sentiment'], cannot
 autodetect index. Will continue without an index.
To pick an index column, use the index= keyword; to 
silence this warning use index=False.
  RuntimeWarning,
Sentiment Analysis Query
95.80851052730421% overlap with 1533536 rows (RAPIDS denominator)
95.90632204766145% overlap with 1531972 rows (Spark denominator)
Correctness Assertion True
Please install gspread and oauth2client               to use Google Sheets automation
```

**Comment by Christian8491** [comment]
Updated and tested again at sf 1K on the dgx201. It shows Correctness Assertion True (using the negativeSentiment.txt file just to test)

(tpcx) christianc@rl-dgx2-d17-u05-rapids-dgx201:~/tpcx-bb/tpcx_bb/queries/q18$ python tpcx_bb_query_18_sql.py --config_file=/home/christianc/yaml_dir/validation_1k_config.yaml
Using default arguments
Connected to tcp://10.150.162.27:8786
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.protobuf.UnsafeUtil (file:/home/christianc/miniconda3/envs/tpcx/lib/blazingsql-algebra.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of com.google.protobuf.UnsafeUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
BlazingContext ready
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/cudf/core/column/string.py:683: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/cudf/core/column/string.py:683: UserWarning: `n` parameter is not supported when                 `pat` and `repl` are list-like inputs
  `pat` and `repl` are list-like inputs"
/home/christianc/miniconda3/envs/tpcx/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py:720: RuntimeWarning: Multiple sorted columns found ['s_name', 'sentiment'], cannot
 autodetect index. Will continue without an index.
To pick an index column, use the index= keyword; to 
silence this warning use index=False.
  RuntimeWarning,
Sentiment Analysis Query
95.80851052730421% overlap with 1533536 rows (RAPIDS denominator)
95.90632204766145% overlap with 1531972 rows (Spark denominator)
Correctness Assertion True
Please install gspread and oauth2client               to use Google Sheets automation

**Pull request merged**